modules/currency-l0/src/main/scala/org/tessellation/currency/l0/snapshot/services/StateChannelBinarySender.scala

```diff
@@ -0,0 +1,195 @@
+package org.tessellation.currency.l0.snapshot.services
+
+import cats.Applicative
+import cats.data.NonEmptyList
+import cats.effect.{Async, Ref, Temporal}
+import cats.syntax.all._
+
+import org.tessellation.currency.l0.node.IdentifierStorage
+import org.tessellation.node.shared.domain.cluster.storage.L0ClusterStorage
+import org.tessellation.node.shared.domain.snapshot.storage.LastSnapshotStorage
+import org.tessellation.node.shared.domain.statechannel.StateChannelValidator.StateChannelValidationError
+import org.tessellation.node.shared.http.p2p.clients.StateChannelSnapshotClient
+import org.tessellation.schema.{GlobalIncrementalSnapshot, GlobalSnapshotInfo, SnapshotOrdinal}
+import org.tessellation.security.{Hashed, Hasher}
+import org.tessellation.statechannel.StateChannelSnapshotBinary
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.types.all.NonNegLong
+import eu.timepit.refined.types.numeric.PosLong
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+import retry._
+
+case class TrackedBinary(
+  binary: Hashed[StateChannelSnapshotBinary],
+  sentAtOrdinal: SnapshotOrdinal
+)
+
+trait StateChannelBinarySender[F[_]] {
+  def processPending: F[Unit]
+  def clearPending: F[Unit]
+  def confirm(globalSnapshot: GlobalIncrementalSnapshot): F[Unit]
+  def process(binaryHashed: Hashed[StateChannelSnapshotBinary]): F[Unit]
+}
+
+object StateChannelBinarySender {
+  def make[F[_]: Async: Hasher: Temporal](
+    identifierStorage: IdentifierStorage[F],
+    globalL0ClusterStorage: L0ClusterStorage[F],
+    lastGlobalSnapshotStorage: LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo],
+    stateChannelSnapshotClient: StateChannelSnapshotClient[F]
+  ): F[StateChannelBinarySender[F]] = (
+    Ref.of[F, List[TrackedBinary]](List.empty),
+    Ref.of[F, PosLong](PosLong.MinValue),
+    Ref.of[F, NonNegLong](NonNegLong.MinValue),
+    Ref.of[F, Boolean](false)
+  ).mapN(make[F](_, _, _, _, identifierStorage, globalL0ClusterStorage, lastGlobalSnapshotStorage, stateChannelSnapshotClient))
+
+  def make[F[_]: Async: Hasher: Temporal](
+    pendingR: Ref[F, List[TrackedBinary]],
+    expectedOrdinalsToConfirmR: Ref[F, PosLong],
```

Comment:
It's set in one place but never fetched -> effectively seems like not needed.

Reply:
Yeah, now it can be just a pure function as we use it in the same place where we calculate it.

Actually, it is a question if `expectedOrdinalsToConfirm` should be calculated from the confirmation ordinal speed of the current global snapshot or the previous one.

If current -> that `Ref` is not needed and we can calculate it in place
If previous -> it is still needed and we need to `getAndSet`

Reply:
I decided to use the current values and in some cases make a fallback to prev (hence the Ref is still needed)